### PR TITLE
perf: scope updates of issue and redeem transactions

### PR DESCRIPTION
**Pull request by anton-karlovskiy**

### CODE REVIEW

File: src/common/live-data/redeem-transaction.watcher.ts [hunk 1]
```diff
@@ -1,18 +1,27 @@
...
 
-    const { address, polkaBtcLoaded, bitcoinHeight } = store.getState().general;
+    const {
+      address,
+      polkaBtcLoaded,
+      bitcoinHeight
+    } = store.getState().general;
     if (!address || !polkaBtcLoaded) return;
 
     const parachainHeight = await window.polkaBTC.system.getCurrentBlockNumber();
```
Comment (nud3l) [inline src/common/live-data/redeem-transaction.watcher.ts:27]: Would it be worth to wrap that in a `Promise.all()`?

**Review by nud3l — COMMENTED** [review]
This PR solves the scoping part. I'm happy to merge like this.

There is another thing that needs doing for the SG though. Right now, this updates the issue and redeem requests are updated every minute. However, we know when we would need to update them by filtering events.

The `window.polkaBTC` API is a was connection. We can use that to listen to events like: https://polkadot.js.org/docs/api/examples/promise/system-events
We are using this construction in polkabtc-js to check if certain events are issued to decide if transactions were successful by checking if the event log contains a certain event: https://github.com/interlay/polkabtc-js/blob/master/src/utils/transaction.ts#L103

So instead of just "blindly" every minute, what I would like to do is to listen to those events and if we detect that there is a event with `section === issue` (see https://github.com/interlay/polkabtc-js/blob/master/src/parachain/issue.ts#L141) for this specific `address` that we currently have, only then we would query the DB to update the issue/redeem requests.

File: src/common/live-data/redeem-transaction.watcher.ts (hunk 1 above)
Comment (anton-karlovskiy) [reply src/common/live-data/redeem-transaction.watcher.ts:27]: Nice catch!

Comment (anton-karlovskiy) [reply src/common/live-data/redeem-transaction.watcher.ts]: Done!

**Comment by nud3l** [comment]
Thanks for the updates! I was thinking a bit more about this and my proposal is to just roll with the interval updates for now as the load on the DB can easily be scaled and calls to the DB are cheap and fast.

The wss connection to the parachain on the other hand is relatively expensive, so instead of making a fix now that includes all the events, it is probably worth thinking about this in more detail and clarify with other teams what the preferred way of connection is.

What do you think?

**Comment by anton-karlovskiy** [comment]
@nud3l 
The implementation is done. But I ran into an issue.
When any issue event occurs, we catch that event considering my account address. So as a result, when any of my issue events (`CancelIssue`, `ExecuteIssue`, or `RequestIssue`) occurs, we fetch issue transactions and the next workflow stays the same (updating the redux store and reflecting the updated state on the UI).
But the problem is that the newly fetched issue transactions do not include the latest issue event.
I looked into the logs and tested for a while and I don't think there's a problem with the implementation.

> only then we would query the DB to update the issue/redeem requests

As far as I'm concerned, there seems to be a race condition.
Although we catch that event and fetch issue transactions from the DB, DB has not been updated with the latest issue event by that time.

race-condition

FYI: I commented out the console logs for now but you can open them and try.

**Comment by anton-karlovskiy** [comment]
> Thanks for the updates! I was thinking a bit more about this and my proposal is to just roll with the interval updates for now as the load on the DB can easily be scaled and calls to the DB are cheap and fast.
> 
> The wss connection to the parachain on the other hand is relatively expensive, so instead of making a fix now that includes all the events, it is probably worth thinking about this in more detail and clarify with other teams what the preferred way of connection is.
> 
> What do you think?

The newly proposed way using the wss connection to the parachain actually sounded sophisticated.
But as I mentioned at https://github.com/interlay/polkabtc-ui/pull/169#issuecomment-802784212, there are some problems to resolve.

> it is probably worth thinking about this in more detail and clarify with other teams what the preferred way of connection is.

Sounds good to me.
Actually in my opinion `setInterval` is not expensive in the Web. But considering some edge cases, it might be problematic.
So I'd suggest using nested `setTimeout`.

Let me explain via our Frontend doc.

**Comment by nud3l** [comment]
You are right, there is a race condition that I did not take into consideration.

So that is happening is the following:

1. Transaction is made by either the user or a vault
2. Event is fired on the parachain
3. In parallel:
  - UI is notified of the event and queries the DB
  - Stats is notified of the event and adds it to the DB
4. If the event is not yet placed in the DB, the UI will not update

So I guess instead of subscribing to the chain, we would actually need to subscribe to the DB right? That way we would prevent the race condition.

**Comment by nud3l** [comment]
One workaround could be to query for a specific request instead of all requests for that user. The event is always specific to a single request, so instead of querying all user requests, stats could provide a Promise to resolve a specific request.

@thecodingrobot @theodorebugnet what do you guys think about that?

**Comment by nud3l** [comment]
> > Thanks for the updates! I was thinking a bit more about this and my proposal is to just roll with the interval updates for now as the load on the DB can easily be scaled and calls to the DB are cheap and fast.
> > The wss connection to the parachain on the other hand is relatively expensive, so instead of making a fix now that includes all the events, it is probably worth thinking about this in more detail and clarify with other teams what the preferred way of connection is.
> > What do you think?
> 
> The newly proposed way using the wss connection to the parachain actually sounded sophisticated.
> But as I mentioned at [#169 (comment)](https://github.com/interlay/polkabtc-ui/pull/169#issuecomment-802784212), there are some problems to resolve.
> 
> > it is probably worth thinking about this in more detail and clarify with other teams what the preferred way of connection is.
> 
> Sounds good to me.
> Actually, in my opinion, `setInterval` is not expensive on the Web. But considering some edge cases, it might be problematic like queueing API calls.
> So I'd suggest using nested `setTimeout`.
> 
> Let me explain via our Frontend doc with other notes I collected in the process of perfomance profiling.

Ok, then let's go with the `setTimeout` and potentially we can add a subscription to events in `stats` so we prevent race conditions in the future.

**Comment by anton-karlovskiy** [comment]
@nud3l 
For now, we can roll back to the version of `serInterval` and later I'm going to create another PR with `nested setTimeout`.

**Comment by theodorebugnet** [comment]
Sounds good to me. I agree that overall, polling the database might easily be better than subscribing to the parachain, and in the future subscribing to the database would be a better optimisation.
I think PostgreSQL might even have native events that can trigger external code that can perform side-effects (though I'm not sure, @thecodingrobot can probably confirm/deny). Failing that, an API can easily be added by hand to provide that.

**Comment by anton-karlovskiy** [comment]
@nud3l 
I reverted to the version of `setInterval`.
Could you please review and merge if it looks good?
Thank you.

**Review by nud3l — APPROVED** [review]

**Comment by nud3l** [comment]
LGTM!

**Pull request merged**